### PR TITLE
RST-3172 Refactor catkin_virtualenv to allow locking dependencies

### PR DESCRIPTION
**Pull request by paulbovbel**
Please start with the README changes https://github.com/locusrobotics/catkin_virtualenv/pull/55/files#diff-04c6e90faac2675aa89e2176d2eec7d8R117

Ripped out all the 'requirements' munging and replaced it with `pip-tools` - this lets us easily manage a `requirements.txt` lockfile, built from `requirements.in` specifications. `pip-tools` is the backend used by the `pipenv` project, however we don't need all the virtualenv management `pipenv` provides.

Removed the dh_virtualenv vendoring, and moved all the lifted code into a `relocate.py` that keeps the dh_virtualenv copyright header.

### CODE REVIEW

File: catkin_virtualenv/cmake/catkin_generate_virtualenv.cmake [hunk 1]
```diff
@@ -17,42 +17,55 @@
...
+  if(NOT DEFINED ARG_LOCK_FILE)
+    set(ARG_LOCK_FILE "${CMAKE_BINARY_DIR}/requirements.txt")
+    message(WARNING "Please define a LOCK_FILE relative to your sources, and commit together to prevent dependency drift.")
+  else()
+    set(ARG_LOCK_FILE "${CMAKE_SOURCE_DIR}/${ARG_LOCK_FILE}")
   endif()
 
   if (NOT DEFINED ARG_EXTRA_PIP_ARGS)
-    set(ARG_EXTRA_PIP_ARGS "-qq")
+    set(ARG_EXTRA_PIP_ARGS "-qq" "--retries 10" "--timeout 30")
```
Comment (paulbovbel) [inline catkin_virtualenv/cmake/catkin_generate_virtualenv.cmake:53]: @garyservin, why did we need to build in `retry` when pip had retries all along...

File: catkin_virtualenv/src/catkin_virtualenv/relocate.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+def find_script_files(venv_dir):
+    """Find list of files containing python shebangs in the bin directory. """
+    command = [
+        'grep', '-l', '-r',
+        '-e', r'^#!.*bin/\(env \)\?{0}'.format(_PYTHON_INTERPRETERS_REGEX),
+        '-e', r"^'''exec.*bin/{0}".format(_PYTHON_INTERPRETERS_REGEX),
+        os.path.join(venv_dir, 'bin')]
+    grep_proc = subprocess.Popen(command, stdout=subprocess.PIPE)
+    files, _ = grep_proc.communicate()
+    return set(f for f in files.decode('utf-8').strip().split('\n') if f)
```
Comment (ablakey) [inline catkin_virtualenv/src/catkin_virtualenv/relocate.py:40]: Don't forget set comprehensions exist.
```
return {f for f in files.decode('utf-8').strip().split('\n') if f}
```

Comment (paulbovbel) [reply catkin_virtualenv/src/catkin_virtualenv/relocate.py]: Done!

**Comment by paulbovbel** [comment]
Migrated https://github.com/locusrobotics/aiorospy/pull/30, will add to docs as example

**Comment by paulbovbel** [comment]
Apologized for all the force-push noise, retagging reviewers.

Long story short, my initial plan intended to use pip-compile to aggregate _all_ of a catkin package's pip dependencies (including those of its catkin ancestors) into a `requirements.txt` lock file. This completely wrecked the development model.

1. `packageB` depends on `packageA` and both use catkin_virtualenv to declare some loose input requirement `requirements.in`.
2. `packageB` contains a locked  `requirements.txt`  of all `package(A|B)` dependencies.
3. `package A` has a new feature added which requires adding `library1` to `requirements.in`.
4.  `packageB` is broken, because it imports code from `packageA`, however its lock file hasn't been updated with `library1`. 

In other worlds, this is fine, because `packageB` depends on `packageA=0.1.0`, however in our world, we can't be bothered to keep track of inter-library versioning, and `packageB` depends on `packageA=latest`. So it goes.

So, the new way is, from step 2:

2. `packageB` contains a locked `requirements.txt` of only `packageB` dependencies, and wants  on `packageA` to declare its own fully locked `requirements.txt`
3. When a `library1` is added to `packageA`'s requirements.txt, its `requirements.txt` must be regenerated to include it.

The locked dependencies are now distributed amongst the whole catkin package chain, which isn't what I initially hoped for, but it's still better than the status quo.

**Comment by MJohnson459** [comment]
This seems a reasonable approach. I can see a potential issue when you have more than one dependency i.e. A depends on B and C, B depends on numpy 1.16 and C depends on numpy 1.18. We can probably manage that in practice but it will make it difficult to update the version of any dependency in the future.

Also if we start versioning and tracking dependencies for our own code the initial issues, do all of the original problems disappear? That might be something we should start doing anyway.

**Comment by paulbovbel** [comment]
> Also if we start versioning and tracking dependencies for our own code the initial issues, do all of the original problems disappear? That might be something we should start doing anyway.

So, we track dependencies, but we don't typically version them (within catkin). We typically rely on Ubuntu and our apt mirror to lock system dependencies, the rosdistro to lock upstream catkin depdencies, and then let our own internal packages roll with `devel`, until it's time to pin a release.

I believe what you're pitching in terms of versioning dependencies is, that hypothetically speaking, we could create a system where locus_navigation would depend on locus_costmap==0.1 and locus_mapping would depend on locus_costmap==0.2. It's certainly technically possible.

I feel like this would be a bit of a nightmare for us to manage, and it's a lot more expedient to say 'there shall be only one version of costmap used by all downstream projects, and any API breakages in costmap will be fixed downstream as necessary'.

Part of me deeply regrets the freedom we've established via catkin_virtualenv to mix-and-match dependency versions amongst projects, and I wish we had gone with a method to install one blessed set of versioned python libraries system-wide.

**Pull request merged**